### DISCUSSION TITLE
Brainstorm ideas for our programming language design.

### DISCUSSION THREAD

**Discussion by thehale** [jhale1805/codeable#1] [Ideas]
Use this discussion thread to brainstorm ideas for our language!

**Comment by thehale** [comment]
Some goofy ideas to get things started.
- An emoji-lang like [emojicode](https://github.com/emojicode/emojicode) but with an easier to read syntax.
- A language where all the keywords are the names of our Software Engineering Faculty.
    - Other ASU themed languages could be fun.
- A language that requires a physical device to write
   - e.g. a Rubik's Cube, Game Controller, etc.
- Writing all identifiers in a different spoken language (e.g. Spanish, Hindi, Ukranian, etc.)

**Comment by JakeHresh** [comment]
Whatever the keywords look like, I think we should play to our strengths and make a language that involves defining solutions in terms of algorithms, that is, HOW to solve the problem, not WHAT the solution is. In other words, not like Prolog.

**Reply by thehale** [reply]
You're in luck! Prolog deliberately doesn't have an assignment operator, which is a required language feature for our language.

![image](https://user-images.githubusercontent.com/47901316/159112604-e581e44d-89c5-4806-8d02-b76297d55c7e.png)

**Comment by JakeHresh** [comment]
Would a WYSIWYG be on the table?

**Reply by thehale** [reply]
Do you have an example of what a WYSIWYG type of language would look like?

**Reply by thehale** [reply]
Here's an idea on this front, a language that deliberately avoids using special symbols to avoid confusing people with no experience with programming languages at all.

Functions could look like this
```
<function_name> needs <arg0> [with <argN>]*
<internal_operations>
answer equals <return_value>
```

Calling a function looks like this
```
<function_name> of <arg0> [with <argN>]*
```

An example program (the quadratic formula) could look like this
```
addition needs augend with addend
answer equals augend plus addend

subtraction needs minuend with subtrahend
answer equals minuend minus subtrahend

multiplication needs multiplicand with multiplier
answer equals multiplicand times multiplier

division needs dividend with divisor
answer equals dividend divided-by divisor

remainder needs dividend with divisor
answer equals dividend modulo divisor

exponentiation needs base with exponent
answer equals base raised-to exponent

square needs number
answer equals exponentiation of number with 2

square-root needs number
answer equals exponentiation of number with 0.5

quadratic-formula needs a with b with c
b-squared equals square of b
four-a equals multiplication of 4 with a
four-a-c equals multiplication of four-a with c
discriminant equals subtraction of b-squared with four-a-c
sqrt-discriminant equals square-root of discriminant
negative-b equals subtraction of 0 with b
numerator equals addition of negative-b with sqrt-discriminant
denominator equals multiplication of 2 with a
answer equals division of numerator with denominator
```

All the variables are deliberately immutable, and no operator precedence is supported (which effectively makes this a functional language), which is why the `quadratic-formula` function is so long.

**Reply by thehale** [reply]
Ideas for loop syntaxes (requirement #4)

```
for <item_name> from <collection>
<internal_operations>
repeat
```

```
while <boolean_condition>
<internal_operations>
repeat
```

**Reply by thehale** [reply]
Ideas for `if-then-else` syntaxes (requirement #3b)

```
if <boolean_condition>
<internal_operations>
otherwise if <boolean_condition>
<internal_operations>
otherwise 
<internal_operations>
```
> TODO How should we end the `if-then-else` statement?
> We could require an empty line after the block, but then it could be confused with a function.
> I also don't know if we want indentation, since that's a new concept to beginning programmers...

Inline ternary (requirement #3a)
```
answer equals <value> if <boolean_condition> otherwise <value>
```

**Reply by thehale** [reply]
Comments would be written as follows
```
fyi <comment>
```

**Reply by 07sainishanth** [reply]
are we going to use `is equals to` for comparing `a == b`?
also, do you have anything in mind for function definitions?
what paradigm is our language going to be?
